connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,266 @@
+[Early-data-specification] section.
+
+#### Early data specification
+
+The early data message is encoded in the "protobuf2" syntax as shown in the
+following. The protobuf definition is an extension to [handshake-payload]. The
+existing byte array early data (the "data" field) will be replaced by a
+structured NoiseExtensions protobuf message. The supported muxers and selected
+muxer are populated in the "stream_muxers" field. The details of the early
+data message can be find in [Noise-handshake-payload]
+
+The muxers are ordered by preference, with the most prefered muxer at the
+beginning.
+
+```protobuf
+syntax = "proto2";
+
+message NoiseExtensions {
+    repeated bytes webtransport_certhashes = 1;
+    repeated string stream_muxers = 2; 
+}
+
+message NoiseHandshakePayload {
+  optional bytes identity_key = 1;
+  optional bytes identity_sig = 2;
+  optional NoiseExtensions extensions = 4;
+}
+```
+
+## Cross version support
+
+The improved muxer selection approach MUST be interoperable with pervious
+libp2p versions which do not support this improved approach.
+
+### TLS case
+
+In the current version of libp2p, the ALPN extension field is populated with a
+key "libp2p". By appending the key of "libp2p" to the end of the supported
+muxer list, the TLS handshaking process is not broken when peers run different
+versions of libp2p, because the minimum overlap of the peer's supported muxer
+sets is always satisfied. When one peer runs the old version and the other peer
+runs the version that supports this feature, the negotiated protocol is
+"libp2p".
+
+In the case "libp2p" is the result of TLS ALPN, an empty result MUST be
+returned to the upgrade process to indicate that no muxer was selected. And the
+upgrade process MUST fall back to the multistream-selection protocol to
+to negotiate the muxer to be selected.
+
+### Noise case
+
+The existing version of libp2p Noise handshake carries empty early data. When a
+version that supports this feature talks to an older version which does not
+support this feature, the muxer selection process on the new version runs
+against an empty string and will return empty muxer selection result.
+
+In the case an empty muxer selection result is returned, the upgrade process
+MUST fall back to the multistream-selection protocol to select the muxer.
+
+## Security
+
+The muxer list carried in TLS ALPN extension field is part of the ClientHello
+message which is not encrypted. This feature will expose the supported muxers
+in plain text, but this is not a weakening of securiy posture. In the fuure
+when [ECH] is ready the muxer info can be protected too.
+
+The early data in Noise handshake is only sent afer the peers establish a
+shared key, in the second and third handshake messages in the XX pattern. So
+the early data is encrypted and the muxer info carried over is protected.
+These is no security weakening in this case either.
+
+
+## Protocol coupling
```

Comment:
That also seems to belong in the introduction, and can probably be shortened to a single sentence.

Reply:
done